### PR TITLE
entity linking to pytorch and reduce ram and gpu

### PR DESCRIPTION
**Pull request by dmitrijeuseew**

### CODE REVIEW

File: deeppavlov/models/entity_extraction/entity_detection_parser.py [hunk 1]
```diff
@@ -45,10 +45,9 @@ def __call__(self, questions: List[str]) -> List[str]:
 class EntityDetectionParser(Component):
     """This class parses probabilities of tokens to be a token from the entity substring."""
 
-    def __init__(self, entity_tags: List[str], type_tag: str, o_tag: str, tags_file: str, ignore_points: bool = False,
+    def __init__(self, o_tag: str, tags_file: str, entity_tags: List[str] = None, ignore_points: bool = False,
```
Comment (IgnatovFedor) [inline deeppavlov/models/entity_extraction/entity_detection_parser.py:48]: update docstring, plus entity_tags is Optional[List[str]]

**Review by IgnatovFedor — CHANGES_REQUESTED** [review]

File: deeppavlov/models/kbqa/rel_ranking_infer.py [hunk 2]
```diff
@@ -12,42 +12,74 @@
...
+    """Class for ranking of paths in subgraph"""
 
     def __init__(self, load_path: str,
                  rel_q2name_filename: str,
-                 ranker: RelRanker,
-                 rels_to_leave: int = 15,
-                 batch_size: int = 100, **kwargs):
-
+                 ranker=None,
+                 bert_preprocessor=None,
```
Comment (IgnatovFedor) [inline deeppavlov/models/kbqa/rel_ranking_infer.py:36]: typing

File: deeppavlov/models/kbqa/rel_ranking_infer.py [hunk 3]
```diff
@@ -56,35 +88,139 @@ def load(self) -> None:
...
+                            candidate_rels_str = " # ".join([self.rel_q2name[candidate_rel] \
+                                                             for candidate_rel in candidate_rels if
+                                                             candidate_rel in self.rel_q2name])
+                        if candidate_rels_str:
+                            questions_batch.append(question)
+                            rels_batch.append(candidate_rels)
+                            rels_labels_batch.append(candidate_rels_str)
+                            answers_batch.append(candidate_answer)
+                            entities_batch.append(candidate_entities)
+                            confidences_batch.append(candidate_confidence)
```
Comment (IgnatovFedor) [inline deeppavlov/models/kbqa/rel_ranking_infer.py:132]: Could candidate_ans_and_rels be False and candidate_rels_str be True? In this case we will reference variables before assignment

File: deeppavlov/models/kbqa/template_matcher.py [hunk 4]
```diff
@@ -114,9 +115,11 @@ def __call__(self, question: str, entities_from_ner: List[str]) -> \
                         query_type = template["template_type"]
                         entity_types = template.get("entity_types", [])
                         template_answer = template.get("template_answer", "")
+                        answer_types = template.get("answer_types", [])
                         min_length = cur_len
 
-        return entities, types, relations, relation_dirs, query_type, entity_types, template_answer, template_found
+        return entities, types, relations, relation_dirs, query_type, entity_types, template_answer, answer_types, \
```
Comment (IgnatovFedor) [inline deeppavlov/models/kbqa/template_matcher.py:121]: doesn't match expected response type

File: deeppavlov/models/kbqa/type_define.py [hunk 5]
```diff
@@ -0,0 +1,138 @@
+import pickle
+from typing import List
+import spacy
+import pymorphy2
+from nltk.corpus import stopwords
+from deeppavlov.core.commands.utils import expand_path
+from deeppavlov.core.common.registry import register
```
Comment (IgnatovFedor) [inline deeppavlov/models/kbqa/type_define.py:7]: add license and sort and group imports

File: deeppavlov/models/kbqa/wiki_parser.py [hunk 6]
```diff
@@ -31,7 +31,22 @@
...
+                     "label": "http://wl",
+                     "alias": "http://wal",
+                     "description": "http://wd",
+                     "rels": {"direct": "http://wpd",
+                              "no_type": "http://wp",
+                              "statement": "http://wps",
+                              "qualifier": "http://wpq"
+                              },
+                     "statement": "http://ws"
+                 },
```
Comment (IgnatovFedor) [inline deeppavlov/models/kbqa/wiki_parser.py:47]: default argument is mutable, replace default value by None

File: deeppavlov/models/kbqa/wiki_parser_online.py [hunk 7]
```diff
@@ -1,107 +0,0 @@
...
-from requests.exceptions import ConnectionError, ConnectTimeout, ReadTimeout
-
-from deeppavlov import __version__ as dp_version
-from deeppavlov.core.common.registry import register
-
-log = getLogger(__name__)
-
-
-@register('wiki_parser_online')
-class WikiParserOnline:
```
Comment (IgnatovFedor) [inline deeppavlov/models/kbqa/wiki_parser_online.py:29]: remove from the registry.json

File: deeppavlov/models/kbqa/rel_ranking_bert_infer.py [hunk 8]
```diff
@@ -1,190 +0,0 @@
...
-from deeppavlov.core.common.file import load_pickle
-from deeppavlov.models.ranking.rel_ranker import RelRanker
-from deeppavlov.models.kbqa.wiki_parser import WikiParser
-from deeppavlov.models.preprocessors.bert_preprocessor import BertPreprocessor
-from deeppavlov.models.kbqa.sentence_answer import sentence_answer
-
-log = getLogger(__name__)
-
-
-@register('rel_ranking_bert_infer')
```
Comment (IgnatovFedor) [inline deeppavlov/models/kbqa/rel_ranking_bert_infer.py:30]: mentioned in the deeppavlov/models/kbqa/sentence_answer.py line 31 and in the docs/features/models/kbqa.rst line 263

File: deeppavlov/models/kbqa/query_generator_online.py [hunk 9]
```diff
@@ -1,190 +0,0 @@
...
-from deeppavlov.models.kbqa.rel_ranking_infer import RelRankerInfer
-from deeppavlov.models.kbqa.rel_ranking_bert_infer import RelRankerBertInfer
-from deeppavlov.models.kbqa.utils import \
-    extract_year, extract_number, make_combs, fill_online_query
-from deeppavlov.models.kbqa.query_generator_base import QueryGeneratorBase
-
-log = getLogger(__name__)
-
-
-@register('query_generator_online')
```
Comment (IgnatovFedor) [inline deeppavlov/models/kbqa/query_generator_online.py:33]: still in registry.json

File: deeppavlov/models/preprocessors/torch_transformers_preprocessor.py [hunk 10]
```diff
@@ -181,6 +174,50 @@ def __call__(self, texts_a: List[str], texts_b: Optional[List[str]] = None) -> U
         return input_features
 
 
+@register('torch_transformers_entity_ranker_preprocessor')
```
Comment (IgnatovFedor) [inline deeppavlov/models/preprocessors/torch_transformers_preprocessor.py:177]: remove @register: component didn't used from the configuration file (it's even doesn't added to the registry)

File: deeppavlov/models/preprocessors/torch_transformers_preprocessor.py [hunk 11]
```diff
@@ -181,6 +174,50 @@ def __call__(self, texts_a: List[str], texts_b: Optional[List[str]] = None) -> U
...
+class TorchTransformersEntityRankerPreprocessor(Component):
```
Comment (IgnatovFedor) [inline deeppavlov/models/preprocessors/torch_transformers_preprocessor.py:178]: add docstring with the minimal description

File: deeppavlov/models/preprocessors/torch_transformers_preprocessor.py [hunk 12]
```diff
@@ -274,6 +311,63 @@ def __call__(self, texts_a: List[str], texts_b: Optional[List[str]] = None) -> U
             return input_features
 
 
+@register('rel_ranking_preprocessor')
+class RelRankingPreprocessor(Component):
```
Comment (IgnatovFedor) [inline deeppavlov/models/preprocessors/torch_transformers_preprocessor.py:315]: add docstring with the minimal description

File: deeppavlov/models/torch_bert/torch_transformers_el_ranker.py [hunk 13]
```diff
@@ -0,0 +1,353 @@
...
+import numpy as np
+from torch import Tensor
+# from apex import amp
+
+from deeppavlov.core.commands.utils import expand_path
+from transformers import AutoConfig, AutoTokenizer, AutoModel
+from deeppavlov.core.common.errors import ConfigError
+from deeppavlov.core.common.registry import register
+from deeppavlov.core.models.torch_model import TorchModel
+from deeppavlov.models.preprocessors.torch_transformers_preprocessor import TorchTransformersEntityRankerPreprocessor
```
Comment (IgnatovFedor) [inline deeppavlov/models/torch_bert/torch_transformers_el_ranker.py:17]:
```suggestion
from logging import getLogger
from pathlib import Path
from typing import List, Optional, Dict, Tuple, Union, Any

import numpy as np
import torch
import torch.nn as nn
import torch.nn.functional as F
from torch import Tensor
from transformers import AutoConfig, AutoTokenizer, AutoModel

from deeppavlov.core.commands.utils import expand_path
from deeppavlov.core.common.errors import ConfigError
from deeppavlov.core.common.registry import register
from deeppavlov.core.models.torch_model import TorchModel
from deeppavlov.models.preprocessors.torch_transformers_preprocessor import TorchTransformersEntityRankerPreprocessor
```

File: deeppavlov/models/torch_bert/torch_transformers_el_ranker.py [hunk 14]
```diff
@@ -0,0 +1,353 @@
+from pathlib import Path
```
Comment (IgnatovFedor) [inline deeppavlov/models/torch_bert/torch_transformers_el_ranker.py:1]: add license

File: deeppavlov/models/torch_bert/torch_transformers_el_ranker.py [hunk 15]
```diff
@@ -0,0 +1,353 @@
...
+from transformers import AutoConfig, AutoTokenizer, AutoModel
+from deeppavlov.core.common.errors import ConfigError
+from deeppavlov.core.common.registry import register
+from deeppavlov.core.models.torch_model import TorchModel
+from deeppavlov.models.preprocessors.torch_transformers_preprocessor import TorchTransformersEntityRankerPreprocessor
+
+log = getLogger(__name__)
+
+
+@register('torch_transformers_el_ranker')
```
Comment (IgnatovFedor) [inline deeppavlov/models/torch_bert/torch_transformers_el_ranker.py:22]: doesn't used as component in configuration files, remove @register

File: deeppavlov/models/torch_bert/torch_transformers_el_ranker.py [hunk 16]
```diff
@@ -0,0 +1,353 @@
...
+class TorchTransformersElRanker(TorchModel):
```
Comment (IgnatovFedor) [inline deeppavlov/models/torch_bert/torch_transformers_el_ranker.py:23]: add docstring with minimal information

File: deeppavlov/models/torch_bert/torch_transformers_el_ranker.py [hunk 17]
```diff
@@ -0,0 +1,353 @@
...
+    def __init__(
+            self,
+            model_name: str,
+            encoder_save_path: str,
+            bilinear_save_path: str,
+            pretrained_bert: str = None,
+            bert_config_file: Optional[str] = None,
+            criterion: str = "CrossEntropyLoss",
+            optimizer: str = "AdamW",
+            optimizer_parameters: Dict = {"lr": 5e-5, "weight_decay": 0.01, "eps": 1e-6},
```
Comment (IgnatovFedor) [inline deeppavlov/models/torch_bert/torch_transformers_el_ranker.py:34]: replace default value with None

File: deeppavlov/models/torch_bert/torch_transformers_el_ranker.py [hunk 18]
```diff
@@ -0,0 +1,353 @@
...
+            return_probas: bool = False,
+            attention_probs_keep_prob: Optional[float] = None,
+            hidden_keep_prob: Optional[float] = None,
+            clip_norm: Optional[float] = None,
+            threshold: Optional[float] = None,
```
Comment (IgnatovFedor) [inline deeppavlov/models/torch_bert/torch_transformers_el_ranker.py:39]: doesn't used

File: deeppavlov/models/torch_bert/torch_transformers_el_ranker.py [hunk 19]
```diff
@@ -0,0 +1,353 @@
...
+                pred = torch.argmax(softmax_scores, dim=1).cpu().numpy()
+
+        return pred
+
+    def siamese_ranking_el_model(self, **kwargs) -> nn.Module:
+        return SiameseBertElModel(
+            pretrained_bert=self.pretrained_bert,
+            encoder_save_path=self.encoder_save_path,
+            bilinear_save_path=self.bilinear_save_path,
+            bert_tokenizer_config_file=self.pretrained_bert,
```
Comment (IgnatovFedor) [inline deeppavlov/models/torch_bert/torch_transformers_el_ranker.py:118]: unexpected argument

File: deeppavlov/models/torch_bert/torch_transformers_el_ranker.py [hunk 20]
```diff
@@ -0,0 +1,353 @@
...
+        with torch.no_grad():
+            softmax_scores = self.model(**_input)
+            if self.return_probas:
+                pred = softmax_scores
+            else:
+                pred = torch.argmax(softmax_scores, dim=1).cpu().numpy()
+
+        return pred
+
+    def siamese_ranking_el_model(self, **kwargs) -> nn.Module:
```
Comment (IgnatovFedor) [inline deeppavlov/models/torch_bert/torch_transformers_el_ranker.py:113]: why you need kwargs? And why object initialization is standalone method?

File: deeppavlov/models/torch_bert/torch_transformers_el_ranker.py [hunk 21]
```diff
@@ -0,0 +1,353 @@
...
+            "optimizer_state_dict": self.optimizer.state_dict(),
+            "epochs_done": self.epochs_done
+        }, weights_path)
+        self.model.to(self.device)
+        self.model.save()
+
+
+class TextEncoder(nn.Module):
+    def __init__(self, pretrained_bert: str = None,
+                 bert_tokenizer_config_file: str = None,
```
Comment (IgnatovFedor) [inline deeppavlov/models/torch_bert/torch_transformers_el_ranker.py:140]: unused argument

File: deeppavlov/models/torch_bert/torch_transformers_el_ranker.py [hunk 22]
```diff
@@ -0,0 +1,353 @@
...
+                 bert_config_file: str = None,
+                 resize: bool = False,
```
Comment (IgnatovFedor) [inline deeppavlov/models/torch_bert/torch_transformers_el_ranker.py:142]: unused argument

File: deeppavlov/models/torch_bert/torch_transformers_el_ranker.py [hunk 23]
```diff
@@ -0,0 +1,353 @@
...
+        torch.save({
+            "model_state_dict": self.model.cpu().state_dict(),
+            "optimizer_state_dict": self.optimizer.state_dict(),
+            "epochs_done": self.epochs_done
+        }, weights_path)
+        self.model.to(self.device)
+        self.model.save()
+
+
+class TextEncoder(nn.Module):
```
Comment (IgnatovFedor) [inline deeppavlov/models/torch_bert/torch_transformers_el_ranker.py:138]: add docstring with the minimal information

File: deeppavlov/models/torch_bert/torch_transformers_el_ranker.py [hunk 24]
```diff
@@ -0,0 +1,353 @@
...
+
+        elif self.bert_config_file and Path(self.bert_config_file).is_file():
+            self.config = AutoConfig.from_json_file(str(expand_path(self.bert_config_file)))
+            self.encoder = AutoModel.from_config(config=self.bert_config)
+        else:
+            raise ConfigError("No pre-trained BERT model is given.")
+        self.encoder.to(self.device)
+
+
+class BilinearRanking(nn.Module):
```
Comment (IgnatovFedor) [inline deeppavlov/models/torch_bert/torch_transformers_el_ranker.py:192]: add docstring with minimal information

File: deeppavlov/models/torch_bert/torch_transformers_el_ranker.py [hunk 25]
```diff
@@ -0,0 +1,353 @@
...
+        b1 = text1.view(-1, self.emb_size // self.block_size, self.block_size)
+        b2 = text2.view(-1, self.emb_size // self.block_size, self.block_size)
+        bl = (b1.unsqueeze(3) * b2.unsqueeze(2)).view(-1, self.emb_size * self.block_size)
+        logits = self.bilinear(bl)
+        softmax_logits = self.softmax(logits)
+        log_softmax = F.log_softmax(logits, dim=-1)
+        return softmax_logits, log_softmax
+
+
+class SiameseBertElModel(nn.Module):
```
Comment (IgnatovFedor) [inline deeppavlov/models/torch_bert/torch_transformers_el_ranker.py:211]: add docstring with minimal information

File: deeppavlov/models/torch_bert/torch_transformers_el_ranker.py [hunk 26]
```diff
@@ -0,0 +1,353 @@
...
+        torch.save({"model_state_dict": self.encoder.cpu().state_dict()}, encoder_weights_path)
+        bilinear_weights_path = expand_path(self.bilinear_save_path).with_suffix(f".pth.tar")
+        log.info(f"Saving bilinear weights to {bilinear_weights_path}.")
+        torch.save({"model_state_dict": self.bilinear_ranker.cpu().state_dict()}, bilinear_weights_path)
+        self.encoder.to(self.device)
+        self.bilinear_ranker.to(self.device)
+
+
+@register('torch_transformers_entity_ranker_infer')
+class TorchTransformersEntityRankerInfer:
```
Comment (IgnatovFedor) [inline deeppavlov/models/torch_bert/torch_transformers_el_ranker.py:273]: add docstring with minimal description

File: deeppavlov/requirements/spacy.txt [hunk 27]
```diff
@@ -1 +1 @@
-spacy>=2.2.3,<=2.3.3
+spacy==3.2.4
```
Comment (IgnatovFedor) [inline deeppavlov/requirements/spacy.txt:1]: could be a range of versions?

File: docs/features/models/kbqa.rst [hunk 28]
```diff
@@ -20,18 +20,13 @@ The question answerer:
...
-* :config:`kbqa_rus <kbqa/kbqa_cq_rus.json>` - for answering complex questions over Wikidata in Russian,
+* :config:`kbqa_cq_ru <kbqa/kbqa_cq_ru.json>` - for answering complex questions over Wikidata in Russian,
 
-* :config:`kbqa_cq_online <kbqa/kbqa_cq_online.json>` - for answering complex questions in English over Wikidata using Wikidata Query Service.
-
-The first two models are very similar to each other, and they allow you to deploy them together with local copy of Wikidata on-premises or in the cloud. The third model is lightweight as it allows you to skip downloading entire Wikidata and use the existing Wikidata APIs instead. 
-
-.. note:: 
-   We recommend you to use the lightweight model for quick experiments as well as academic research, and full models in production to avoid dependencies on the public Wikidata APIs.
+These configs use local Wikidata dump in hdt format (3.7 Gb on disk). :config:`kbqa_cq_en <kbqa/kbqa_cq_en.json>` requires 3.5 Gb RAM and 4.3 Gb GPU, :config:`kbqa_cq_ru <kbqa/kbqa_cq_ru.json>` - 6.9 Gb RAM and 6.5 Gb GPU.
```
Comment (IgnatovFedor) [inline docs/features/models/kbqa.rst:29]: maybe add this to the table?

File: deeppavlov/configs/classifiers/query_pr.json [hunk 29]
```diff
@@ -1,115 +1,146 @@
...
-        "url": "http://files.deeppavlov.ai/deeppavlov_data/bert/cased_L-12_H-768_A-12.zip",
-        "subdir": "{DOWNLOADS_PATH}/bert_models_kbqa"
+        "url": "http://files.deeppavlov.ai/kbqa/wikidata/query_prediction_eng.tar.gz",
+        "subdir": "{MODELS_PATH}/classifiers/query_prediction_eng"
       },
       {
-        "url": "http://files.deeppavlov.ai/kbqa/models/query_prediction.tar.gz",
-        "subdir": "{MODELS_PATH}/classifiers/query_prediction"
+        "url": "http://files.deeppavlov.ai/kbqa/wikidata/query_prediction_eng.pickle",
+        "subdir": "{DOWNLOADS_PATH}/query_prediction"
```
[the remark this answers is not in the document]
Comment (dmitrijeuseew) [reply deeppavlov/configs/classifiers/query_pr.json:143]: query_prediction_eng.tar.gz is the model checkpoint and query_prediction_eng.pickle is the dataset. I think that if the model checkpoint is updated, it should be uploaded to share separately.

File: deeppavlov/configs/entity_extraction/entity_linking_en.json [hunk 30]
```diff
@@ -0,0 +1,66 @@
...
+    },
+    "download": [
+      {
+        "url": "http://files.deeppavlov.ai/deeppavlov_data/entity_linking/el_db_eng.tar.gz",
+        "subdir": "{DOWNLOADS_PATH}/entity_linking_eng"
+      },
+      {
+        "url": "http://files.deeppavlov.ai/deeppavlov_data/entity_linking/el_ranker_eng.tar.gz",
+        "subdir": "{MODELS_PATH}/entity_linking_eng"
+      },
```
[the remark this answers is not in the document]
Comment (dmitrijeuseew) [reply deeppavlov/configs/entity_extraction/entity_linking_en.json:55]: el_db_eng.tar.gz is wikidata database and el_ranker_eng.tar.gz is the model checkpoint, I think that if database will be updated, it should be separately uploaded to share.

File: deeppavlov/configs/kbqa/kbqa_cq_en.json [hunk 31]
```diff
@@ -0,0 +1,108 @@
...
+        "rel_ranker": "#rel_r_inf",
+        "wiki_parser": "#wiki_p",
+        "load_path": "{DOWNLOADS_PATH}/wikidata",
+        "rank_rels_filename_1": "rels_0.txt",
+        "rank_rels_filename_2": "rels_1.txt",
+        "sparql_queries_filename": "{DOWNLOADS_PATH}/wikidata/sparql_queries.json",
+        "entities_to_leave": 5,
+        "rels_to_leave": 10,
+        "return_answers": true,
+        "in": ["x", "x", "template_type", "f_entity_substr", "f_tags", "answer_types"],
```
[the remark this answers is not in the document]
Comment (dmitrijeuseew) [reply deeppavlov/configs/kbqa/kbqa_cq_en.json:72]: In kbqa_cq_ru.json the first "x" is initial question, the second "x_sanitized" is the question reformulated using syntax parser.

File: deeppavlov/configs/kbqa/kbqa_cq_en.json [hunk 32]
```diff
@@ -0,0 +1,108 @@
+{
+  "chainer": {
+    "in": ["x_init"],
```
[the remark this answers is not in the document]
Comment (dmitrijeuseew) [reply deeppavlov/configs/kbqa/kbqa_cq_en.json:3]: replaced with "x"

File: deeppavlov/configs/kbqa/kbqa_cq_en.json [hunk 33]
```diff
@@ -0,0 +1,108 @@
...
+    },
+    "download": [
+      {
+        "url": "http://files.deeppavlov.ai/kbqa/wikidata/sparql_queries.json",
+        "subdir": "{DOWNLOADS_PATH}/wikidata"
+      },
+      {
+        "url": "http://files.deeppavlov.ai/kbqa/wikidata/rels.tar.gz",
+        "subdir": "{DOWNLOADS_PATH}/wikidata"
+      },
```
[the remark this answers is not in the document]
Comment (dmitrijeuseew) [reply deeppavlov/configs/kbqa/kbqa_cq_en.json:93]: separate files allow easier uploaded, for example, updated sparql_queries.json

File: deeppavlov/models/entity_extraction/entity_detection_parser.py [hunk 34]
```diff
@@ -60,30 +59,31 @@ def __init__(self, entity_tags: List[str], type_tag: str, o_tag: str, tags_file:
...
         self.o_tag = o_tag
         self.ignore_points = ignore_points
         self.return_entities_with_tags = return_entities_with_tags
         self.thres_proba = thres_proba
         self.tag_ind_dict = {}
         with open(str(expand_path(tags_file))) as fl:
             tags = [line.split('\t')[0] for line in fl.readlines()]
+            if self.entity_tags is None:
+                self.entity_tags = list(
+                    {tag.split('-')[1] for tag in tags if len(tag.split('-')) > 1}.difference({self.o_tag}))
```
[the remark this answers is not in the document]
Comment (dmitrijeuseew) [reply deeppavlov/models/entity_extraction/entity_detection_parser.py:71]: For example, "B-PER" and "I-PER" will be converted into "PER". "O" - tag when split by "-" will have only one element, for this case len(tag.split('-')) > 1.

File: deeppavlov/models/kbqa/query_generator.py [hunk 35]
```diff
@@ -74,18 +75,27 @@ def __call__(self, question_batch: List[str],
...
+            if template_type == "-1":
+                template_type = "7"
+            candidate_outputs, template_answer, templates_nums = \
+                self.find_candidate_answers(question, question_sanitized, template_type, entities_from_ner,
+                                            entity_tags_list, answer_types)
             candidate_outputs_batch.append(candidate_outputs)
             template_answers_batch.append(template_answer)
+            templates_nums_batch.append(templates_nums)
+        qg_tm2 = time.time()
+        log.debug(f"--------query generator time {qg_tm2 - qg_tm1}")
```
[the remark this answers is not in the document]
Comment (dmitrijeuseew) [reply deeppavlov/models/kbqa/query_generator.py:98]: Deleted

File: deeppavlov/models/kbqa/query_generator.py [hunk 36]
```diff
@@ -95,12 +105,13 @@ def __call__(self, question_batch: List[str],
             return answers
         else:
             log.debug(f"(__call__)candidate_outputs_batch: {[output[:5] for output in candidate_outputs_batch]}")
-            return candidate_outputs_batch
+            return candidate_outputs_batch, entities_from_ner_batch, template_answers_batch
```
[the remark this answers is not in the document]
Comment (dmitrijeuseew) [reply deeppavlov/models/kbqa/query_generator.py:108]: It was a debug parameter, deleted.

File: deeppavlov/models/kbqa/query_generator.py [hunk 37]
```diff
@@ -126,14 +137,18 @@ def query_parser(self, question: str, query_info: Dict[str, str],
                              search_source != "do_not_rank"]
         log.debug(f"(query_parser)rel_directions: {triplet_info_list}")
         entity_ids = [entity[:self.entities_to_leave] for entity in entity_ids]
+        rel_tm1 = time.time()
```
[the remark this answers is not in the document]
Comment (dmitrijeuseew) [reply deeppavlov/models/kbqa/query_generator.py:140]: debug info, deleted

File: deeppavlov/models/kbqa/query_generator.py [hunk 38]
```diff
@@ -203,18 +226,29 @@ def query_parser(self, question: str, query_info: Dict[str, str],
...
             else:
-                candidate_outputs = [output[1:] for output in candidate_outputs]
-        log.debug(f"(query_parser)loop time: {datetime.datetime.now() - start_time}")
+                candidate_outputs = [{"entities": f_entities,
+                                      "relations": f_relations,
+                                      "answers": f_answers,
+                                      "rel_conf": f_rel_conf
+                                      } for f_entities, *f_relations, f_answers, f_rel_conf in candidate_outputs]
+        query_tm2 = time.time()
+        log.debug(f"--------queries execution time: {query_tm2 - query_tm1}")
```
[the remark this answers is not in the document]
Comment (dmitrijeuseew) [reply deeppavlov/models/kbqa/query_generator.py:251]: debug code, deleted

File: deeppavlov/models/kbqa/query_generator_base.py [hunk 39]
```diff
@@ -119,89 +127,104 @@ def find_candidate_answers(self, question: str,
...
 
         if entities_from_template or types_from_template:
             if rels_from_template[0][0] == "PHOW":
                 how_to_content = self.find_answer_wikihow(entities_from_template[0])
                 candidate_outputs = [["PHOW", how_to_content, 1.0]]
             else:
-                entity_ids = self.get_entity_ids(entities_from_template, "entities", template_found, question,
-                                                 entity_types)
-                type_ids = self.get_entity_ids(types_from_template, "types")
+                el_tm1 = time.time()
```
[the remark this answers is not in the document]
Comment (dmitrijeuseew) [reply deeppavlov/models/kbqa/query_generator_base.py:148]: debug code, deleted

File: deeppavlov/models/preprocessors/torch_transformers_preprocessor.py (hunk 10 above)
Comment (dmitrijeuseew) [reply deeppavlov/models/preprocessors/torch_transformers_preprocessor.py]: added config for training where this component is used

File: deeppavlov/models/torch_bert/torch_transformers_el_ranker.py (hunk 15 above)
Comment (dmitrijeuseew) [reply deeppavlov/models/torch_bert/torch_transformers_el_ranker.py:22]: added config configs/ranking/el_ranking_distilbert_eng.json, where this component is used

File: deeppavlov/models/torch_bert/torch_transformers_el_ranker.py (hunk 20 above)
Comment (dmitrijeuseew) [reply deeppavlov/models/torch_bert/torch_transformers_el_ranker.py:153]: I used the example from https://github.com/deepmipt/DeepPavlov/blob/191e3603cd06267a877cd711e3ffc11232524496/deeppavlov/models/classifiers/torch_classification_model.py#L182

File: deeppavlov/models/torch_bert/torch_transformers_el_ranker.py [hunk 40]
```diff
@@ -0,0 +1,432 @@
...
+class SiameseBertElModel(nn.Module):
+    """Class with model for ranking of entities by context and description
+    Args:
+        emb_size: entity embedding size
+        block_size: size of block in bilinear layer
+        encoder_save_path: path to save the encoder checkpoint
+        bilinear_save_path: path to save bilinear layer checkpoint
+        pretrained_bert: pretrained Bert checkpoint path or key title (e.g. "bert-base-uncased")
+        bert_config_file: path to Bert configuration file, or None, if `pretrained_bert` is a string name
+        device: `cpu` or `cuda` device to use
```
Comment (IgnatovFedor) [inline deeppavlov/models/torch_bert/torch_transformers_el_ranker.py:274]: doesn't handled correctly. By default 'gpu' value raises RuntimeError

File: deeppavlov/models/torch_bert/torch_transformers_el_ranker.py [hunk 41]
```diff
@@ -0,0 +1,432 @@
...
+    """Class for obtaining the BERT output for CLS-token and special entity token
+    Args:
+        pretrained_bert: pretrained Bert checkpoint path or key title (e.g. "bert-base-uncased")
+        bert_config_file: path to Bert configuration file, or None, if `pretrained_bert` is a string name
+        device: `cpu` or `cuda` device to use
+    """
+
+    def __init__(self, pretrained_bert: str = None,
+                 bert_config_file: str = None,
+                 device: str = "gpu"
```
Comment (IgnatovFedor) [inline deeppavlov/models/torch_bert/torch_transformers_el_ranker.py:190]: default value doesn't handled correctly

**Review by IgnatovFedor — CHANGES_REQUESTED** [review]

File: deeppavlov/models/kbqa/query_generator.py [hunk 42]
```diff
@@ -178,22 +180,30 @@ def query_parser(self, question: str, query_info: Dict[str, str],
         if self.wiki_file_format == "pickle":
             total_entities_list = list(itertools.chain.from_iterable(selected_entity_ids)) + \
                                   list(itertools.chain.from_iterable(selected_type_ids))
-            parse_res = self.wiki_parser(["parse_triplets"], [total_entities_list])
+            try:
+                self.wiki_parser(["parse_triplets"], [total_entities_list])
```
[the remark this answers is not in the document]
Comment (IgnatovFedor) [reply deeppavlov/models/kbqa/query_generator.py:183]: remove wiki_file_format == "pickle"

**Review by IgnatovFedor — APPROVED** [review]

**Pull request merged**